### ISSUE TITLE
[BUG] Stop recomputing transition matrices when they haven't changed

### ISSUE THREAD

**Issue by bredelings** [revbayes/revbayes#232]
Currently RevBayes recomputes the transition probability matrix for a branch every time we compute conditional likelihoods for a node.

https://github.com/revbayes/revbayes/blob/master/src/core/distributions/phylogenetics/substitution/PhyloCTMCSiteHomogeneous.h#L218=

https://github.com/revbayes/revbayes/blob/master/src/core/distributions/phylogenetics/substitution/PhyloCTMCSiteHomogeneous.h#L280=

There is a patch to fix this on the origin/dev_jsigao branch:

https://github.com/revbayes/revbayes/commit/ac3e10e1a29cf8a679318c4ec0961023949fd538

@jsigao says on slack that this leads to 50-200% speedups, but needs to be thoroughly tested.  I imagine that even larger speedups could be seen for codon models.

This issue is tracking this problem and discussion of how/when to solve it.

**Comment by bredelings** [comment]
The indexing formulae for getting the transition probability matrix for (active, node, mixture_component) seems like it should be implemented as an inline function.  The same goes for conditional likelihoods too, of course.  The only justification for manually computing the indexes into the 1-dimensional array seems to be speed ... am I missing something?

@hoehna 
@jsigao

**Comment by bredelings** [comment]
I'm also curious whether we could switch to something like
```
vector<shared_ptr< const vector<TransitionProbabilityMatrix>>> active;
vector<shared_ptr<const vector<TransitionProbabilityMatrix>>> inactive;
```
with the following rules:
1. If the `inactive` vector has a non-zero length, then we can switch back to the old states by doing
```
active = inactive;
inactive.clear();
```
2. For the active vector, then we have a valid cached version of the transition probability matrices for node `node` if `active[node]` is not null.  If `active[node]` is not null, then the P matrices the `node` are given by the vector `*active[node]`.  In this  case, maybe we don't need to store a separate indicator for whether the node is dirty?

3. Since we have a `const` flag on each entry, the active vector isn't going to accidentally modify the inactive vector.

**Comment by bredelings** [comment]
Also, I think that maybe we SHOULD call  `this->updateTransitionProbabilities( node_index );` every time that we compute conditional likelihoods.

The difference is that we should make that function do something like

```
const vector<TransitionProbabilityMatrix>& getTransitionProbabilities( node_index)
{
    if (not active[node_index])
         active[node_index] = std::make_shared<TransitionProbabilitymatrix>( computeTransitionProbabilities(node_index) );
    return *active[node_index]
}
```

In other words, we should lazily compute the TransitionProbability matrix vector on demand.  If there is a cached value available, then we can return that without doing any work.

**Comment by bredelings** [comment]
This should be fixed by #277
